Embedding Cludo assets and searches, along with settings. KBHBIB-76

rasben opened the pull request:
I've tried to break things up as much as I could into descriptive commits, but the really long story short is:

- Creating a custom settings form, that allows the editors to choose which Cludo searches should be embedded
<img width="864" height="1007" alt="Screenshot 2025-09-15 at 14 16 52" src="https://github.com/user-attachments/assets/67b3edc5-637b-4ddc-8435-a9a2ef25be45" />
- These cludo searches are defined as a `CludoProfile`, containing the information that Cludo needs to embed forms/autosuggestions/results. Some of this information needs to be sent along to Cludo using Javascript.
- If a Profile is enabled, we'll redirect if an existing view already exists (e.g. `views.articles.all`)
- A new paragraph has also been introduced, that allows an editor to insert a search form related to a `CludoProfile` on a page:

<img width="833" height="576" alt="Screenshot 2025-09-15 at 14 24 12" src="https://github.com/user-attachments/assets/b1867b6e-7c47-4494-ae4c-67f0f820673c" />

- Finally, we also introduce a way to call the Cludo API directly, serverside.
Right now, we only need it for when we're on the material search page, and want to tell if there are any relevant results, and how many.


<img width="759" height="269" alt="Screenshot 2025-09-15 at 14 32 39" src="https://github.com/user-attachments/assets/f32dfd87-30d2-4243-ac49-691fbcacd620" />

In the future, we will also use that functionality to tell Cludo when content is updated, and get it indexed faster.


(https://github.com/kdb/kdb_cludo/pull/5)

rasben committed:
Create CludoProfile class, containing context for a CludoEngine. KBHBIB-76

Embedding a Cludo search requires sending various properties along to
Cludo. The CludoProfile class will allow us to easily set up contexts in
the future.

rasben committed:
Setup module settings and base for search page. KBHBIB-76

- A settings form for kdb_cludo, allowing editors to enable Cludo
profiles as search pages.
This commit introduces no available profiles, but sets the base.

- An abstract controller base, used for creating Cludo search pages, and
loading the correct Cludo assets for embeddings.

rasben committed:
Create 'main' CludoSearch engine/page and settings. KBHBIB-76

rasben committed:
Make sure kdb_cludo config is ignored by config_ignore_auto.

Adding `hook_install` and `hook_uninstall` functionality, to make sure
that `*kdb_cludo*` is toggled in `config_ignore_auto` when the module is
enabled/disabled.

rasben committed:
When on a Cludo search page, replace standard material search header. KBHBIB-76

This allows the user to see and alter their search query.
For this to work, we also had to update the hook weight within the
module, to make sure our logic happens after the dpl code that inserts
the material search header.

rasben committed:
Add a link to the editorial web search, on the material search page. KBHBIB-76

When on the material search, the default DPL setup creates a "Look for
this query on the website" link, if we find any results.
If the `main` CludoProfile is enabled, we want the link to instead lead
here, along with the query.

To begin with, the link is always shown, without any context if any
results are available.
In the future, once Cludo API has been implemented, we will add this.

rasben committed:
Handles redirecting to relevant Cludo search pages, when necessary.

The module allows editors to enable/disable Cludo search pages.

When they are enabled, we want the core search pages to redirect, and if
they are not enabled, we want the Cludo pages to redirect to the core
search (or set access denied, if none)

rasben committed:
Introduce all the Cludo search pages/profiles, available.

rasben committed:
Introduce Cludo-search paragraph, embedding search on any page.

rasben committed:
Expand Cludo paragraph to allow editor to set breadcrumb context.

The breadcrumb - if set - is shown, once the user is navigated onto the
relevant search page.

rasben committed:
Create service for calling Cludo's API - and expand settings form.

rasben committed:
Set preview-count on search-link that is shown on material search.

rasben committed:
Load Cludo script assets using dynamic customerID.

rasben committed:
Fix Cludo form styling. KBHBIB-76

rasben committed:
Only replace material search on `main` cludo search. KBHBIB-86

Also removes the `default_input` parameter, as it does not actually
matter - Cludo sets it by itself, using the URL.

rasben committed:
Introduce simple/advanced SERPs for Cludo search pages.

Most of the Cludo searches have filters, that need to be sent along as
part of the template.
We'll add an option that can be set dynamically, along with a template.

Ideally, we would want to do the alternative template as a theme hook
suggestion, but apparantly this logic doesnt work in a standalone
modules template folder.

rasben committed:
Throw exceptions / log problems when CludoAPI is not set up yet.

rasben committed:
Add 'is wide' option for Cludo paragraph. KBHBIB-88

rasben committed:
Deny access to Cludo SERP if disabled.

rasben committed:
Nicer Cludo search URLs.

rasben committed:
Simplify breadcrumb-logic for Cludo Search paragraph. KBHBIB-87

rasben committed:
Re-align form input with paragraph texts. KBHBIB-93

rasben committed:
Set correct classes and IDs for different search pages. KBHBIB-90

rasben committed:
Invalidate caches when breadcrumb URL parameter changes. KBHBIB-92

rasben committed:
Show page title as part of search breadcrumb. KBHBIB-92

rasben committed:
Set correct viewRoute for events. KBHBIB-94

rasben committed:
Allow kdb_cludo paragraph to be selected in articles and pages.

xendk commented on assets/cludo-search-init.js:
```diff
@@ -0,0 +1,30 @@
+
+  // The searchResults script from Cludo uses the main engineId in the URL,
+  // rather than the currently selected profiles.
+  const mainEngineId = '14490';
```
You should decide whether you want to make this generic and reusable, or just hardcode it to KBH. Both options have merit, but the module should be consistent in which way it goes. If it goes for configurable, this engineId needs to be configurable too.

rasben replied:
These IDs will never change, that's why I've put it there - rather than putting another abstraction layer of config ontop

I think I prefer to keep it like this, but your comment is appreciated and I'll keep it in mind for future additions.

xendk commented on kdb_cludo.install:
```diff
@@ -0,0 +1,67 @@
+ * @return string
+ *   The feedback message.
+ */
+function _kdb_cludo_config_auto_unignore_entites(array $entities): string {
```
What's wrong with `_dpl_update_config_auto_unignore_entites()`?

rasben replied:
`_dpl_update_config_auto_unignore_entites()` happens in `dpl_update.install`, which means i can't access the function here.
Atleast, I didnt seem to be able to - even after setting `dpl_update` as a module dependency in `kdb_cludo`

xendk commented on kdb_cludo.module:
```diff
@@ -167,3 +174,247 @@ function kdb_cludo_page_attachments(array &$page): void {
+    return;
+  }
+
+  $variables['#cache']['tags'][] = 'kdb_cludo';
```
Rather than adding this tag and invalidating it in the settings form submission, add the relevant config entities cache tags. Then Drupal will handle it for you.

rasben replied:
I get your point, but the cache tags are added dynamically, and I feel like this works just as well.

xendk commented on kdb_cludo.module:
```diff
@@ -167,3 +174,247 @@ function kdb_cludo_page_attachments(array &$page): void {
+    return;
+  }
+
+  $url = $profile->getCludoUrl();
```
This preprocess function spends over ten lines of code on just trying to figure out if it has an active profile or not. It shouldn't be its job. There ought to be a method on `CludoProfileService` that does it for it.

rasben replied:
I agree with your sentiment, but I don't see which parts of this preprocess that does that.
Can you give me an example of one of the lines that you think should be moved to `CludoProfileService`?

xendk commented on kdb_cludo.module:
```diff
@@ -167,3 +174,247 @@ function kdb_cludo_page_attachments(array &$page): void {
+    $hasWebResults = TRUE;
+    $totalResults = NULL;
+
+    if ($apiService->isAvailable()) {
```
And here the preprocess function has to determine if the service is available, and know the format of request and responses from the Cludo API. It shouldn't, it should just call `$apiService->getNemResults(string $input): ?int`, which returns the number or `NULL` if the service isn't available.

xendk commented on src/Controller/CludoSearchBase.php:
```diff
@@ -0,0 +1,131 @@
+    $breadcrumb_node_id = $request->query->get('breadcrumb');
+
+    if ($breadcrumb_node_id) {
+      ;
```
Stray `;`

rasben replied:
weird that phpcs or phpstan didnt care about that :S

xendk commented on src/Controller/CludoSearchBase.php:
```diff
@@ -0,0 +1,131 @@
+
+    if ($this->profile->cludoType === 'help') {
+      $jsSettings['searchInputSelectors'] = ['#cludo-search-input-help'];
+
```
Extra newline.

xendk commented on src/Controller/CludoSearchBase.php:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    $breadcrumb = NULL;
+    $cache_tags = ['kdb_cludo'];
```
See last comment on cache tags.

xendk commented on src/Controller/CludoSearchBase.php:
```diff
@@ -0,0 +1,131 @@
+   *   Render array, for the page.
+   */
+  public function page(Request $request): array {
+    $config = $this->profile->getConfigSettings();
```
Rather than exposing config, add getters to the profile, so users of profiles doesn't have to deal with what is config and what is not.

xendk commented on src/Controller/CludoSearchBase.php:
```diff
@@ -0,0 +1,131 @@
+        }
+
+        $breadcrumb = (new Breadcrumb())->setLinks($breadcrumb_links);
+
```
Another stray newline.

xendk commented on src/CludoProfile.php:
```diff
@@ -0,0 +1,110 @@
+  }
+
+  /**
+   * Child classes override this to provide page-specific settings.
```
Which child classes?

xendk commented on src/Services/CludoProfileService.php:
```diff
@@ -0,0 +1,46 @@
+   */
+  public function getProfiles(): array {
+    return [
+      (new CludoSearchMain())->profile,
```
I do not like this. You're pulling in the controller just to get the profile. Which means you're pulling in everything that `ControllerBase` pulls in. Which is a lot.

You could just define the profiles here, and refactor `CludoSearchBase` into a controller that takes the name of a profile (or dependency inject it through `*.routes.yml` and cut out all the other controllers.

xendk commented on src/Services/CludoApiService.php:
```diff
@@ -0,0 +1,149 @@
+      $this->cludoProfile = $cludoProfile;
+    }
+    else {
+      $this->logger->error('Cludo Profile "main" not found.');
```
This can't happen if the profiles is hardcoded.

rasben committed:
Replace magic config settings with getters for CludoProfile.

rasben committed:
Move CludoProfile initializer to CludoProfileService

Also, standardize the page controllers, to a single base controller that
takes information as a route attribute.

rasben committed:
Introduce and implement a ->getTotalResults for CludoApiService.

rasben committed:
Expose input labels as kdb_cludo setting.
